### PR TITLE
1630 - Add validation for clean job within Estimates by Job Role Pipeline

### PR DESCRIPTION
**Pull request by AchyutRatkanthiwar**
## Description
Trello ticket [#1630](https://trello.com/c/fe7ZcW7f/1630-validation-for-job-roles-clean-job)

Added validations and test validation for clean job within Estimates by Job Role Pipeline.

## Testing
- [x] Unit tests passing
- [x] Successful [Step Function run](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/executions/details/arn:aws:states:eu-west-2:856699698263:execution:1630-jr-clean-val-Ind-CQC-Filled-Post-Estimates-By-Role:4ea1f424-baf6-44d1-a258-c9db193abd60)

## Checklist (delete if not relevant)
- [x] Unit tests added/amended
- [x] Updated CHANGELOG
- [x] Code reviewed by AI
- [x] Moved Trello ticket to PR column

## Reviewer checklist
- [ ] PR solves the Trello ticket requirements
- [ ] Outputs appear reasonable and understandable
- [ ] The overall approach is appropriate and not over-engineered
- [ ] No obvious scalability, performance or interdependency concerns
- [ ] Tests appropriately cover the main behaviour/edge cases
- [ ] Naming and structure are broadly understandable and consistent
- [ ] Docstrings are sufficient for future maintenance
- [ ] Docstrings cover non-obvious behaviour

### CODE REVIEW

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py [hunk 1]
```diff
@@ -40,20 +89,96 @@ def main(bucket_name: str, source_path: str, reports_path: str) -> None:
...
         # complete columns
         .col_vals_not_null(
             [
                 IndCqcColumns.id_per_locationid_import_date,
+                IndCqcColumns.id_per_locationid_import_date_job_role,
                 IndCqcColumns.location_id,
                 IndCqcColumns.cqc_location_import_date,
+                IndCqcColumns.estimate_filled_posts,
+                IndCqcColumns.primary_service_type,
+                IndCqcColumns.main_job_role_clean_labelled,
```
Comment (RoyPriceCS50) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:117]: Add IndCqcColumns.ascwds_filtering_rule

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py [hunk 2]
```diff
@@ -40,20 +89,96 @@ def main(bucket_name: str, source_path: str, reports_path: str) -> None:
...
+        )
+        # categorical
+        .col_vals_in_set(
+            IndCqcColumns.primary_service_type,
+            CatValues.primary_service_type_column_values.categorical_values,
+        )
+        .col_vals_in_set(
+            IndCqcColumns.main_job_role_clean_labelled,
+            CatValues.main_job_role_labels_column_values.categorical_values,
+        )
```
Comment (RoyPriceCS50) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:201]: Add these specially checks

```
.specially(
            vl.is_unique_count_equal(
                IndCqcColumns.primary_service_type,
                CatValues.primary_service_type_column_values.count_of_categorical_values,
            ),
            brief=f"{IndCqcColumns.primary_service_type} should have exactly {CatValues.primary_service_type_column_values.count_of_categorical_values} distinct values",
        )
        .specially(
            vl.is_unique_count_equal(
                IndCqcColumns.main_job_role_clean_labelled,
                CatValues.main_job_role_labels_column_values.count_of_categorical_values,
            ),
            brief=f"{IndCqcColumns.main_job_role_clean_labelled} should have exactly {CatValues.main_job_role_labels_column_values.count_of_categorical_values} distinct values",
        )
```

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py [hunk 3]
```diff
@@ -1,22 +1,59 @@
...
+        IndCqcColumns.cqc_location_import_date: "Date",
+        IndCqcColumns.estimate_filled_posts: "Float32",
+        IndCqcColumns.primary_service_type: str(
+            CategoricalColumnTypes.PrimaryServiceEnumType
+        ),
+        IndCqcColumns.main_job_role_clean_labelled: str(
+            CategoricalColumnTypes.JobRoleEnumType
+        ),
+        IndCqcColumns.ascwds_job_role_counts: "Int16",
+    }
```
Comment (RoyPriceCS50) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:55]: Add IndCqcColumns.ascwds_filtering_rule: CategoricalColumnTypes.JobRoleFilteringRuleCatType

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py [hunk 4]
```diff
@@ -1,22 +1,59 @@
...
-ind_cqc_clean_job_role_cols_to_import = [
+VALIDATION_COLS_TO_IMPORT = [
     IndCqcColumns.id_per_locationid_import_date,
     IndCqcColumns.location_id,
     IndCqcColumns.cqc_location_import_date,
     IndCqcColumns.id_per_locationid_import_date_job_role,
+    IndCqcColumns.estimate_filled_posts,
+    IndCqcColumns.primary_service_type,
+    IndCqcColumns.main_job_role_clean_labelled,
+    IndCqcColumns.ascwds_job_role_counts,
```
Comment (RoyPriceCS50) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:27]: Add IndCqcColumns.ascwds_filtering_rule

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py [hunk 5]
```diff
@@ -40,20 +89,96 @@ def main(bucket_name: str, source_path: str, reports_path: str) -> None:
...
+        .col_vals_expr(
+            expr=(
+                pl.col(IndCqcColumns.ascwds_job_role_counts).sum()
+                < compare_ascwds_job_role_counts_sum
+            ),
+            brief=(
+                f"Total sum of ascwds_job_role_counts in clean job should be less than the total sum of ascwds_job_role_counts in the merge job dataset"
+            ),
+        )
+        # numerical
```
Comment (RoyPriceCS50) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:159]: Add a check here for the ticket bullet point "Check rows where location_job_group_distribution_out_of_bounds is True are null in the ascwds_job_role_counts column".

I think the tickets were made before the clean job was done. The clean job makes a string column to say why counts have been nulled, so think that's what the ticket is supposed to be referencing.

Something like this, but I can't get it to work on my branch for impute validation:

```
.col_vals_expr(
            expr=(
                (
                    (
                        pl.col(IndCqcColumns.job_role_filtering_rule)
                        != pl.lit(JobRoleFilteringRule.populated)
                        & pl.col(IndCqcColumns.ascwds_job_role_counts).is_null()
                    )
                    | (
                        pl.col(IndCqcColumns.job_role_filtering_rule)
                        == pl.lit(JobRoleFilteringRule.populated)
                        & pl.col(IndCqcColumns.ascwds_job_role_counts).is_not_null()
                    )
                )
            ),
            brief="ascwds_job_role_counts must be null where job_role_filtering_rule is not populated",
        )
```

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/test_validate_02_clean.py [hunk 6]
```diff
@@ -1,29 +1,42 @@
...
-            IndCqcColumns.location_id: pl.String,
+            IndCqcColumns.id_per_locationid_import_date: pl.UInt32,
+            IndCqcColumns.location_id: CategoricalColumnTypes.LocationCatType,
             IndCqcColumns.cqc_location_import_date: pl.Date,
-            IndCqcColumns.id_per_locationid_import_date_job_role: pl.String,
+            IndCqcColumns.id_per_locationid_import_date_job_role: pl.UInt32,
+            IndCqcColumns.estimate_filled_posts: pl.Float32,
+            IndCqcColumns.primary_service_type: CategoricalColumnTypes.PrimaryServiceEnumType,
+            IndCqcColumns.main_job_role_clean_labelled: CategoricalColumnTypes.JobRoleEnumType,
+            IndCqcColumns.ascwds_job_role_counts: pl.Int16,
```
Comment (RoyPriceCS50) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/test_validate_02_clean.py:32]: Add IndCqcColumns.ascwds_filtering_rule

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/test_validate_02_clean.py [hunk 7]
```diff
@@ -48,16 +70,25 @@ def test_validation_report_includes_expected_validations(
...
         assertion_types_present = {item["assertion_type"] for item in report_json}
 
-        expected_assertions = {"col_vals_not_null", "rows_distinct"}
+        expected_assertions = {
+            "col_schema_match",
+            "row_count_match",
+            "col_vals_not_null",
+            "rows_distinct",
+            "col_vals_expr",
+            "col_vals_in_set",
```
Comment (RoyPriceCS50) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/test_validate_02_clean.py:90]: Add "specially" (other PR comment)

**Review by RoyPriceCS50 — COMMENTED** [review]
Most feedback is because I think the ticket was made before clean job was done. The clean job makes a string column to show why counts have been nulled and that needs to come into the validation check.

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py (hunk 4 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:27]: Do you mean job_role_filtering_rule?

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py (hunk 3 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:55]: Do you mean job_role_filtering_rule?

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py (hunk 5 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:159]: I will try this out.

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py (hunk 2 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:201]: Added

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py (hunk 3 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:55]: added job_role_filtering_rule

Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:55]: job_role_filtering_rule

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py (hunk 1 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:117]: added job_role_filtering_rule

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/test_validate_02_clean.py (hunk 6 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/test_validate_02_clean.py:32]: added job_role_filtering_rule

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/test_validate_02_clean.py (hunk 7 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/test_validate_02_clean.py:90]: done

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py (hunk 5 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:159]: This worked on mine

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py [hunk 8]
```diff
@@ -40,20 +94,132 @@ def main(bucket_name: str, source_path: str, reports_path: str) -> None:
...
+                    | (
+                        (
+                            pl.col(IndCqcColumns.job_role_filtering_rule)
+                            == pl.lit(JobRoleFilteringRule.populated)
+                        )
+                        & pl.col(IndCqcColumns.ascwds_job_role_counts).is_not_null()
+                    )
+                )
+            ),
+            brief="ascwds_job_role_counts must be null where job_role_filtering_rule is not populated",
```
Comment (MarionJHolloway) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py]: "is not populated" could mean "is null" Could you tweak this to be something like "is a value other than populated"?

**Review by MarionJHolloway — APPROVED** [review]
One small tweak to the wording of a check brief. 

I've approved so it doesn't slow you down :)

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py (hunk 8 above)
Comment (RoyPriceCS50) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py]: Agree.

And I'd put the check in the numeric section rather than index section.

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py (hunk 2 above)
Comment (RoyPriceCS50) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:201]: Put the specially checks in the categorical section, rather that date stuff.

Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py:201]: Done

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py (hunk 8 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_02_clean.py]: Done
